### PR TITLE
feat(agw): benchmark UE MM Context serialization

### PR DESCRIPTION
**Pull request by lionelgo**
## Summary

Quantify and obtain measurements of the memory and time spent on serialization of UE MM context using protobuf library, using a minimal test object

Benchmark performance data such as memory, CPU spent, time during serialization and deserialization.
 
For time benchmarking, std::chrono is used and measurements are flushed to stdout.

For memory benchmarking, [rusage](https://linux.die.net/man/2/getrusage) is used.

## Test Plan

  Benchmarking for AGW is done on the magma dev VM. You just need to bring only this VM up.

Inside magma VM, build mme:

```bash
[VM] cd $MAGMA_ROOT/lte/gateway
[VM] make benchmark_pb
```

  in magma VM
  ```bash
  # N = number of UEs connected
  N=50
  # Measurements are collected from stdout
  sudo service magma@magmad stop
  sudo rm -f /var/lib/redis/dump.rdb /var/opt/magma/redis_dump.rdb stdout.all.txt
  sudo service magma@magmad start
  sudo service magma@mme stop
  sudo service sctpd start
  sudo service magma@mobilityd start
  sudo service magma@pipelined start
  sudo service magma@sessiond start
  sudo rm -f /var/lib/redis/dump.rdb /var/opt/magma/redis_dump.rdb
  sudo cp /var/opt/magma/tmp/mme.conf /usr/local/etc/oai
  sudo cp /var/opt/magma/tmp/spgw.conf /usr/local/etc/oai
  sudo /home/vagrant/build/c/core/oai/oai_mme/mme -p$N | tee stdout.all.txt
```
Here is a script example for extracting values from run (still in magma dev VM):

  ```bash
  cat stdout.all.txt | grep 'TIME PROTOBUF CONVERSION' | cut -d ':' -f2 >> protobuf_conv$N.txt
  cat stdout.all.txt | grep 'TIME PROTOBUF REDIS SERIALIZATION' | cut -d ':' -f2 >> protobuf_redis$N.txt
  cat stdout.all.txt | grep 'Time taken to deserialize contexts' | cut -d ':' -f2 >> protobuf_deserialize_contexts$N.txt
  cat stdout.all.txt | grep 'Time taken to serialize contexts' | cut -d ':' -f2 >> protobuf_serialize_contexts$N.txt
  cat stdout.all.txt | grep 'RUSAGE Contexts serialization'  | grep 'Cpu' | cut -d ':' -f2 >> protobuf_rusage_cpu$N.txt
  cat stdout.all.txt | grep 'RUSAGE Contexts serialization'  | grep 'Maximum resident' | cut -d ':' -f2 >> protobuf_rusage_mrss$N.txt
  cat stdout.all.txt | grep 'RUSAGE Contexts serialization'  | grep 'Page reclaims' | cut -d ':' -f2 >> protobuf_rusage_reclaims$N.txt
  cat stdout.all.txt | grep 'RUSAGE Contexts serialization'  | grep 'Block operations' | cut -d ':' -f2 >> protobuf_rusage_block$N.txt
  cat stdout.all.txt | grep 'RUSAGE Contexts serialization'  | grep 'Context switches ' | cut -d ':' -f2 >> protobuf_rusage_switches$N.txt
```

### CODE REVIEW

File: lte/gateway/c/core/oai/tasks/mme_app/mme_config.c [hunk 1]
```diff
@@ -1843,6 +1843,12 @@ void mme_config_display(mme_config_t* config_pP) {
               config_pP->daylight_saving_time);
   OAILOG_INFO(LOG_CONFIG, "- Run mode .............................: %s\n",
               (RUN_MODE_TEST == config_pP->run_mode) ? "TEST" : "NORMAL");
+#if MME_BENCHMARK
+  if (RUN_MODE_TEST == config_pP->run_mode) {
+    OAILOG_INFO(LOG_CONFIG, "- Test param ...........................: %u\n",
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/tasks/mme_app/mme_config.c]:
```suggestion
    OAILOG_INFO(LOG_CONFIG, "- Benchmark Test param ...........................: %u\n",
```

File: lte/gateway/c/core/oai/tasks/mme_app/experimental/mme_app_serialization.cpp [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+//--C includes -----------------------------------------------------------------
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/tasks/mme_app/experimental/mme_app_serialization.cpp]: this looks like it needs formatting, can we run `clang-format` on this file to fix most of the lint warnings?

**Review by ardzoht — CHANGES_REQUESTED** [review]
minor changes requested

File: lte/gateway/c/core/oai/tasks/mme_app/experimental/mme_app_serialization.cpp (hunk 2 above)
Comment (lionelgo) [reply lte/gateway/c/core/oai/tasks/mme_app/experimental/mme_app_serialization.cpp]: I addressed the lint warnings

**Review by ardzoht — APPROVED** [review]
lgtm

**Pull request merged**